pkg/controllers/managementuser/snapshotbackpopulate/snapshotbackpopulate.go

```diff
@@ -77,58 +89,175 @@ func (h *handler) OnChange(key string, configMap *corev1.ConfigMap) (*corev1.Con
+			_, err = h.etcdSnapshots.Create(&cmGeneratedSnapshot)
+			if err != nil {
+				return configMap, fmt.Errorf("rkecluster %s/%s: error while creating etcd snapshot %s/%s: %w", cluster.Namespace, cluster.Name, cmGeneratedSnapshot.Namespace, cmGeneratedSnapshot.Name, err)
+			}
+		}
+	}
 	return configMap, nil
 }
 
-func (h *handler) configMapToSnapshots(configMap *corev1.ConfigMap) (result []rkev1.ETCDSnapshot, _ error) {
+func (h *handler) configMapToSnapshots(configMap *corev1.ConfigMap, cluster *provv1.Cluster) (result map[string]rkev1.ETCDSnapshot, _ error) {
+	clusterName := cluster.Name
+	clusterNamespace := cluster.Namespace
+	result = map[string]rkev1.ETCDSnapshot{}
 	for k, v := range configMap.Data {
 		file := &snapshotFile{}
 		if err := json.Unmarshal([]byte(v), file); err != nil {
 			logrus.Errorf("invalid non-json value in %s/%s for key %s in cluster %s", configMap.Namespace, configMap.Name, k, h.clusterName)
 			return nil, nil
 		}
+		snapshotName := clusterName + "-" + file.Name
+		// Validate that the corresponding machine for the node exists before creating the snapshot
 		snapshot := rkev1.ETCDSnapshot{
-			Name:      file.Name,
-			Location:  file.Location,
-			Metadata:  file.Metadata,
-			Message:   file.Message,
-			NodeName:  file.NodeName,
-			CreatedAt: file.CreatedAt,
-			Size:      file.Size,
-			Status:    file.Status,
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      snapshotName,
+				Namespace: clusterNamespace,
+				Labels: map[string]string{
+					rke2.ClusterNameLabel: clusterName,
+					rke2.NodeNameLabel:    file.NodeName,
+				},
+				OwnerReferences: []metav1.OwnerReference{},
+			},
+			SnapshotFile: rkev1.ETCDSnapshotFile{
+				Name:      file.Name,
+				Location:  file.Location,
+				Metadata:  file.Metadata,
+				Message:   file.Message,
+				NodeName:  file.NodeName,
+				CreatedAt: file.CreatedAt,
+				Size:      file.Size,
+				Status:    file.Status,
+			},
 		}
+		fileSuffix := "-local"
 		if file.S3 != nil {
-			snapshot.S3 = &rkev1.ETCDSnapshotS3{
+			fileSuffix = "-s3"
+			snapshot.SnapshotFile.S3 = &rkev1.ETCDSnapshotS3{
 				Endpoint:      file.S3.Endpoint,
 				EndpointCA:    file.S3.EndpointCA,
 				SkipSSLVerify: file.S3.SkipSSLVerify,
 				Bucket:        file.S3.Bucket,
 				Region:        file.S3.Region,
 				Folder:        file.S3.Folder,
 			}
+			snapshot.OwnerReferences = append(snapshot.OwnerReferences, metav1.OwnerReference{
+				APIVersion:         cluster.APIVersion,
+				Kind:               cluster.Kind,
+				Name:               cluster.Name,
+				UID:                cluster.UID,
+				Controller:         &[]bool{true}[0],
+				BlockOwnerDeletion: &[]bool{true}[0],
+			})
+		} else {
+			listSuccessful, machine, err := rke2.GetMachineFromNode(h.machineCache, file.NodeName, cluster)
```

Comment:
Are we supposed to only check this in case of local snapshots only?

Reply:
No. We should check this always.

Reply:
To clarify, the owner reference for our snapshot varies based on whether it is an S3 or local backed snapshot 

In the case of S3, it is owned by the cluster
In the case of local, it is owned by the machine that it is is on